### PR TITLE
Blindness rework - damaged eyes are now a stylized simulation of legal blindness

### PR DESCRIPTION
**Pull request by deathride58**
## About the PR
So this started as a half-serious "this kinda resembles IRL blindness, it'd be really cursed to see this ingame" type of thing

But uh...

image

THIS TURNED OUT A **LOT** MORE POPULAR THAN WE EXPECTED

Anyway, we polished the effect up pretty heavily, fixed a few bugs we've noticed with the mechanics surrounding blindness, and even cleaned up the code a bit. This PR description's gonna be lengthy, so bear with us.

Requires space-wizards/RobustToolbox/pull/4783
Requires space-wizards/RobustToolbox/pull/4784

The main meat and potatoes of this PR is the reworked effect for eye damage. This is loosely based off various accounts of various types of legal blindness we've heard over the years, with the most prominent inspiration being that of cataracts, and the second most prominent being neuro-opthalmic injuries. The effect is far from realistic (the realistic route would be bloom + blur heavy enough to make anyone's eyes bleed). Rather, this effect takes the route of being *heavily* stylized. This ultimately gives you the same dilemma that many blind folks in the real world face: you can certainly *try* to see, but in all practicality, you aren't gonna see shit. For those interested in our exact design thoughts: see `Resources/Textures/Shaders/cataracts.swsl`! And for those who don't really care for this wall of text and would rather see the effect in action: scroll down to the Media section!

The actual blindness overlay has also received some attention; total blindness is still very much present. The blindness overlay has been revamped, and is now a soft circle instead of the aliased hard circle previously present. Additionally, this reworked overlay is used in place of the eye damage effect for users who opt in to the new setting introduced by this PR: "Reduce motion of visual effects"; while the eye damage effect is designed to keep on the tame side of things, it's only inevitable that the general waviness might be a bit much for some users!

This PR also makes a few changes on top of that! The most prominent side addition: anyone with sufficient eye damage (enough to reach the maximum strength of the eye damage's effect) will now gain the ability to close their eyes. This allows those with the reworked blind trait to opt to navigate by "feel", and allows anyone near the verge of total blindness to attempt to protect their eyesight.

Glasses now actually do something, as well. Specifically, they now greatly reduce the eye damage visual effect! However, the effectiveness of glasses drops off with higher levels of eye damage, until they eventually make no difference whatsoever.

To put the cherry on top: Because eye damage is now actually debilitating (unlike how it is currently, and unlike how it is in SS13), the amount of damage your eyes can take has been greatly increased. Eyes are now able to take 9 damage, up from 3, and the eye damage effect will max out at 6 damage total. 

... Oh. Right. Another thing: The permanent blindness trait that you can select from character generation is no longer complete blindness. Rather, the blindness trait is now permanent legal blindness, spawning you with 6 eye damage, and preventing you from going any lower than that.

(Regarding the sprites and sounds present: They were made by us! The sounds were made with Dexed synth, ReaJS, and JSFX. The sprites were made in Krita.)

## Why / Balance
The main motivator of why we tinkered with the blindness effect to begin with, and produced the video that prompted the positive reception giving us the motivation to make this whole PR, is because blindness in both SS13 and SS14 is simply borderline unplayable. There has never really been a good middle-ground between being fully blind, and being on the verge of fully blind. And that state of full blindness is one in which it's near-impossible to actually orient yourself, and get a grasp on what's going on in your surroundings. As the primary intent is to make playing blind characters less of a headache, the blind trait now gives characters a permanent minimum of 6 eye damage, instead of spawning the character in with complete blindness.

The new eye damage effect in this PR allows players to have a limited degree of awareness of their surroundings. *Very* limited awareness; past interaction distance, the only thing you can reliably see are lights, but only the vague general direction and general distance of those lights. And while objects are visible a little bit past interaction distance, that visibility is incredibly unreliable due to the distortion effect causing you to see double. This effect is also capable of scaling, with low amounts of eye damage giving you a subtle haze, and gradually ramping up to the point where you can't see clearly past a single tile.

The no-motion variant (for when "reduce motion of visual effects" is enabled in the settings) of the eye damage effect uses the same screenspace effect as the reworked total blindness effect, and is fully capable of scaling. The range has been tuned to roughly match up with what the normal variant is capable of clearly seeing, but completely cuts off past that. This is far more visually clear, but gives far less awareness of your surroundings.

The main thing in this PR that has a mechanical impact is the new ability to close your eyes. This is only available when your eyes are in critical condition. This means you the player are able to actively fight to maintain what's left of your vision, which is especially helpful for the reworked blindness trait. This mechanic allows full blindness to continue existing as a death state of sort for your eyes, as a punishment for continually keeping your eyes exposed to bright sources. This mechanic also allows players to navigate the dark when they're effectively blind, just as they were always able to.

This PR makes a few balance tweaks, as well. Because the new eye damage effect is actually debilitating, the maximum eye damage has been bumped right up. It takes 6 eye damage to max out the eye damage effect and gain the ability to close your eyes, and 9 eye damage total to lose your eyesight completely. Additionally, the examine range modifier for having damaged eyes now scales in a far more reasonable manner!

## Technical details
We've documented as much as possible via comments within our changes, so mostly everything will be self-explanatory

All of the blindness-related systems have been given a code cleanup pass, and now use `Entity<T>` for all of their functions.

VisionCorrectionComponent now no longer gives any `VisionBonus` by default (did you know wearing glasses previously made your eyesight worse as soon as your vision updated?). However, it now has a new field: `CorrectionPower`. This is used as an exponent within blurriness strength calculations for the eye damage effect, and stacks by multiplication.

### Main overview

https://github.com/space-wizards/space-station-14/assets/6356337/5c3dcd56-e068-495c-b337-f6f32f37c93d

### A quick speedrun of all possible states for the eye damage effect
(For demonstration purposes, the temporary blindness from welders has been forced to 1 second. This PR does not affect the temporary blindness effect)

https://github.com/space-wizards/space-station-14/assets/6356337/6fee92e9-86a4-4e26-b82f-4a3dc976dbbe

### And yes, this is compatibility mode compatible!

image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
BlindableSystem, BlindfoldSystem, BlurryVisionSystem, and PermanentBlindnessSystem now use `Entity<T>`.

Additionally, any items defining custom properties for VisionCorrectionComponents will need to be looked at to make sure that their values are sane with the new mechanics.

**Changelog**

:cl: Bhijn and Myr
- add: Blindness has been completely reworked, top to bottom. Eye damage is no longer a mere dimming of the screen, but rather, a hyperstylized (read: not at all realistic) simulation of the symptoms described by those living with various types of real-world blindness.
- add: When your eyes have reached a critical state of damage, you are now capable of simply closing them. 
- add: The settings menu now has a "Reduce motion of visual effects" setting! Currently, the only thing this does is swap the eye damage effect out for a far more straight-forward one akin to the blindness overlays seen across various SS13 servers.
- tweak: The amount of damage that eyes can take has been greatly increased. Total blindness now takes 9 damage, up from 3. Additionally, eyes are considered to be in a critical state once they hit 6 damage.
- tweak: Glasses no longer actively increase the eye damage effect. They now actively reduce the eye damage effect, with their effectiveness dropping off as you take more eye damage, until they eventually do nothing once your eyes are in critical condition.
- tweak: The blind trait is now legal blindness, rather than complete blindness. The blind trait permanently prevents your eyes from being reduced below the critical state threshold.

### CODE REVIEW

**Comment by UbaserB** [comment]
BASED!!

**Comment by Terraspark4941** [comment]
at first i would have thought this looked very jank (and disorienting), but thanks to that accessibility toggle, i ask for merge 
this is cool

**Comment by Hmeister-fake** [comment]
Extraordinarily sick and opens up the prospect of future short-sighted traits and whatnot.

File: Content.Client/Eye/Blinding/BlurryVisionOverlay.cs [hunk 1]
```diff
@@ -44,21 +66,51 @@ protected override bool BeforeDraw(in OverlayDrawArgs args)
...
+            // So the reasonable alternative here is to replace it with a static effect! Specifically, one that replicates the blindness effect seen across most SS13 servers.
+            if (_configManager.GetCVar(CCVars.ReducedMotion))
+            {
+                _circleMaskShader.SetParameter("SCREEN_TEXTURE", ScreenTexture);
+                _circleMaskShader.SetParameter("Zoom", zoom);
+                _circleMaskShader.SetParameter("CircleRadius", NoMotion_Radius / strength);
+
+                worldHandle.UseShader(_circleMaskShader);
+                worldHandle.DrawRect(viewport, Color.White);
+                return;
```
Comment (mirrorcult) [inline Content.Client/Eye/Blinding/BlurryVisionOverlay.cs:100]: useshader(null) after this call?

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**